presto-main/src/main/java/com/facebook/presto/operator/BigintGroupByHash.java

```diff
@@ -339,4 +332,94 @@ private static int calculateMaxFill(int hashSize)
         checkArgument(hashSize > maxFill, "hashSize must be larger than maxFill");
         return maxFill;
     }
+
+    private class AddPageWork
+            implements Work<Void>
+    {
+        private final Block block;
+
+        private int lastPosition;
+
+        public AddPageWork(Block block)
+        {
+            this.block = requireNonNull(block, "block is null");
+        }
+
+        @Override
+        public boolean process()
+        {
+            int positionCount = block.getPositionCount();
+            checkState(lastPosition < positionCount, "position count out of bound");
+
+            // The caller may call process() multiple times even the memory is not available.
+            // We should not proceed if that is the case given we will always add at least one entry to the hash table.
+            if (needRehash() && !tryRehash()) {
+                return false;
+            }
+
+            // Check needRehash() for every position given putIfAbsent() may add an entry but fail to rehash.
+            // If rehash failed in putIfAbsent(), needRehash() will be true and break the loop.
```

Comment:
This comment needs to be more explicit．It should say something like:

putIfAbsent will rehash automatically if rehash is needed, unless there isn't enough memory to do so.